xwiki-pro-macros-api/src/main/java/com/xwiki/macros/userlist/internal/macro/UserListMacro.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.macros.userlist.internal.macro;
+
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.script.ScriptContext;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.macro.AbstractMacro;
+import org.xwiki.rendering.macro.MacroExecutionException;
+import org.xwiki.rendering.transformation.MacroTransformationContext;
+import org.xwiki.script.ScriptContextManager;
+import org.xwiki.template.Template;
+import org.xwiki.template.TemplateManager;
+
+import com.xwiki.macros.userlist.macro.UserListMacroParameters;
+
+@Component
+@Named("userList")
+@Singleton
+public class UserListMacro extends AbstractMacro<UserListMacroParameters>
+{
+    @Inject
+    private TemplateManager templateManager;
+
+    @Inject
+    private ScriptContextManager scriptContextManager;
+
+    /**
+     * Create and initialize the descriptor of the macro.
+     */
+    public UserListMacro()
+    {
+        super("User list", "Displays a custom list of users with their avatar",
+            UserListMacroParameters.class);
+    }
+
+    @Override
+    public List<Block> execute(UserListMacroParameters parameters, String content, MacroTransformationContext context)
+        throws MacroExecutionException
+    {
+        Template customTemplate = this.templateManager.getTemplate("userlist/userlist.vm");
```

Comment:
Why not use ``/html_displayer/userresourcereference/view.vm`` because a displayer supports both **edit** and **view**, and in your case you basically need a picker (edit) for a list of users and a displayer (view) for it.

Moreover, you should actually use the displayer API to render the user list in view mode. Checkout what ``$services.display.html.display(...)`` does under the hood (used in https://github.com/xwiki-contrib/application-ckeditor/blob/master/application-ckeditor-ui/src/main/resources/CKEditor/MacroService.xml#L280 by CKEditor to get the pickers for the macro parameters).

But in order to do this it would be more accurate to use ``List<UserResourceReference>`` as display type for the macro parameter (so that you can pass a list of values when rendering in view mode).

Reply:
The commit d4f70db3 introduces a `UserReferenceList` class used as a macro parameter indeed. It also brings together the edit and view templates in a single common folder, renamed to `userreferencelist` to match the new parameter type.

I will see what I can do with the displayer API.